File: DraftRightWindows/DraftRightWindows/Services/BugReportService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace DraftRightWindows.Services;

/// <summary>
/// Submits bug reports to the DraftRight backend's <c>/bug-reports</c>
/// endpoint as <c>multipart/form-data</c>. Anonymous and authenticated
/// calls are both supported — when a JWT is supplied, the server attaches
/// the user_id automatically.
///
/// API contract: see /tmp/bug-report-contract.md.
/// Source field is hardcoded to <c>"windows"</c>.
/// </summary>
public static class BugReportService
{
    private static readonly HttpClient _http = new() { Timeout = TimeSpan.FromSeconds(60) };
    private const string DefaultBackendUrl = "https://api.draftright.info";

    /// <summary>
    /// Result of a bug report submission. <see cref="Success"/> indicates
    /// HTTP 2xx; <see cref="ErrorMessage"/> carries the failure reason on
    /// non-2xx, network errors, or local exceptions.
    /// </summary>
    public sealed record SubmitResult(bool Success, string? Id = null, string? ErrorMessage = null);

    /// <summary>
    /// POSTs a bug report.
    /// </summary>
    /// <param name="description">Required. Free-text description (min 10 chars enforced by caller).</param>
    /// <param name="screenshotPath">Optional. Path to a PNG/JPEG file on disk (max 5 MB).</param>
    /// <param name="userEmail">Optional. Reporter email — sent only when not signed in.</param>
    /// <param name="authToken">Optional. Bearer JWT — when present, /bug-reports records user_id.</param>
    /// <param name="context">Optional. Extra JSON context (route, plan, last action…).</param>
    /// <param name="backendUrlOverride">Optional. Override base URL (tests / dev). Defaults to settings.</param>
    public static async Task<SubmitResult> SubmitAsync(
        string description,
        string? screenshotPath = null,
        string? userEmail = null,
        string? authToken = null,
        Dictionary<string, object?>? context = null,
        string? backendUrlOverride = null)
    {
        if (string.IsNullOrWhiteSpace(description))
            return new SubmitResult(false, ErrorMessage: "Description is required.");

        // Resolve base URL: explicit override > current Settings > hardcoded default.
        // App.Settings is null in unit/standalone scenarios, so guard before reaching it.
        var baseUrl = backendUrlOverride;
        if (string.IsNullOrWhiteSpace(baseUrl))
        {
            try { baseUrl = App.Settings?.BackendUrl; } catch { baseUrl = null; }
        }
        if (string.IsNullOrWhiteSpace(baseUrl)) baseUrl = DefaultBackendUrl;
        baseUrl = baseUrl!.TrimEnd('/');

        FileStream? screenshotStream = null;
        try
        {
            using var form = new MultipartFormDataContent();

            // Required fields
            form.Add(new StringContent(description), "description");
            form.Add(new StringContent("windows"), "source");

            // App version — assembly version (Package.Current.Id.Version is unreachable
            // in unpackaged WinUI3 builds, which is what DraftRight ships today).
            var asm = System.Reflection.Assembly.GetExecutingAssembly();
            var appVersion = asm.GetName().Version?.ToString() ?? "0.0.0.0";
            form.Add(new StringContent(appVersion), "app_version");

            // OS info — combine OSDescription + arch for triage clarity.
            var arch = System.Runtime.InteropServices.RuntimeInformation.ProcessArchitecture;
            var osInfo =
                $"{System.Runtime.InteropServices.RuntimeInformation.OSDescription} ({Environment.OSVersion.Version}) {arch}";
            form.Add(new StringContent(osInfo), "os_info");

            if (!string.IsNullOrWhiteSpace(userEmail))
                form.Add(new StringContent(userEmail), "user_email");

            // Context JSON — backend accepts a stringified JSON blob in the
            // multipart "context" field and parses it server-side.
            var ctx = context != null
                ? new Dictionary<string, object?>(context)
                : new Dictionary<string, object?>();
            ctx["platform"] = "windows";
            ctx["arch"] = arch.ToString();
            ctx["dotnet"] = Environment.Version.ToString();
            ctx["locale"] = System.Globalization.CultureInfo.CurrentCulture.Name;
            ctx["ts"] = DateTime.UtcNow.ToString("o");
            form.Add(new StringContent(JsonSerializer.Serialize(ctx)), "context");

            // Optional screenshot — open the file as a stream so we don't load
            // the whole image into managed memory.
            if (!string.IsNullOrWhiteSpace(screenshotPath) && File.Exists(screenshotPath))
            {
                var info = new FileInfo(screenshotPath);
                if (info.Length > 5 * 1024 * 1024)
                {
                    return new SubmitResult(false,
                        ErrorMessage: "Screenshot exceeds 5 MB limit.");
                }

                screenshotStream = File.OpenRead(screenshotPath);
                var streamContent = new StreamContent(screenshotStream);
                var ext = Path.GetExtension(screenshotPath).ToLowerInvariant();
                var mime = ext switch
                {
                    ".png" => "image/png",
                    ".jpg" or ".jpeg" => "image/jpeg",
                    _ => "application/octet-stream",
                };
                streamContent.Headers.ContentType = new MediaTypeHeaderValue(mime);
                form.Add(streamContent, "screenshot", Path.GetFileName(screenshotPath));
            }

            using var request = new HttpRequestMessage(HttpMethod.Post, $"{baseUrl}/bug-reports")
            {
                Content = form,
            };

            if (!string.IsNullOrEmpty(authToken))
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", authToken);

            using var response = await _http.SendAsync(request).ConfigureAwait(false);
            var body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);

            if (!response.IsSuccessStatusCode)
            {
                string detail;
                try
                {
                    using var doc = JsonDocument.Parse(body);
                    if (doc.RootElement.TryGetProperty("error", out var err))
                        detail = err.GetString() ?? body;
                    else if (doc.RootElement.TryGetProperty("message", out var msg))
                        detail = msg.GetString() ?? body;
                    else
                        detail = body;
                }
                catch
                {
                    detail = body;
                }
                return new SubmitResult(
                    false,
                    ErrorMessage: $"Server {(int)response.StatusCode}: {detail}");
            }

            string? id = null;
            try
            {
                using var doc = JsonDocument.Parse(body);
                if (doc.RootElement.TryGetProperty("id", out var idProp))
                    id = idProp.GetString();
            }
            catch
            {
                // 2xx but unparseable body — still treat as success.
            }

            return new SubmitResult(true, Id: id);
        }
        catch (TaskCanceledException)
        {
            return new SubmitResult(false, ErrorMessage: "Request timed out. Check your connection and retry.");
        }
        catch (Exception ex)
        {
            return new SubmitResult(false, ErrorMessage: ex.Message);
        }
        finally
        {
            screenshotStream?.Dispose();
        }
    }

    /// <summary>
    /// Minimal description validator used by the dialog (mirrors backend min-length).
    /// </summary>
    public static bool IsDescriptionValid(string? description) =>
        !string.IsNullOrWhiteSpace(description) && description.Trim().Length >= 10;
}

File: DraftRight/Services/BugReportService.swift
import Foundation
import AppKit

/// Submits bug reports to `POST /bug-reports`.
///
/// The endpoint accepts anonymous reports (no auth) and authenticated reports
/// (Bearer JWT). The macOS app passes the Keychain-stored access token when
/// available so reports get associated with the signed-in user.
///
/// Auto-fills `app_version` from `CFBundleShortVersionString`+`CFBundleVersion`
/// and `os_info` from `ProcessInfo.processInfo.operatingSystemVersionString`.
enum BugReportService {
    static let source = "macos"

    enum BugReportError: LocalizedError {
        case invalidURL
        case descriptionTooShort
        case screenshotTooLarge
        case httpError(Int, String)
        case emptyResponse
        case network(Error)

        var errorDescription: String? {
            switch self {
            case .invalidURL: return "Invalid backend URL."
            case .descriptionTooShort: return "Please describe the bug (at least 10 characters)."
            case .screenshotTooLarge: return "Screenshot is larger than 5 MB."
            case .httpError(let code, let body): return "HTTP \(code): \(body)"
            case .emptyResponse: return "No response from server."
            case .network(let err): return err.localizedDescription
            }
        }
    }

    /// Submits a bug report.
    /// - Parameters:
    ///   - description: Free-text user description (min 10 chars after trim).
    ///   - screenshot: Optional PNG/JPEG bytes + filename (e.g. `"shot.png"`).
    ///   - userEmail: Optional override email for anonymous flow.
    ///   - authToken: Optional Bearer JWT — pass non-empty access token when signed in.
    ///   - backendUrl: Base URL e.g. `https://api.draftright.info`.
    /// - Returns: Server-assigned report UUID.
    static func submitReport(
        description: String,
        screenshot: (data: Data, filename: String)? = nil,
        userEmail: String? = nil,
        authToken: String? = nil,
        backendUrl: String
    ) async throws -> UUID {
        let trimmed = description.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count >= 10 else {
            throw BugReportError.descriptionTooShort
        }
        if let shot = screenshot, shot.data.count > 5 * 1024 * 1024 {
            throw BugReportError.screenshotTooLarge
        }

        let base = backendUrl.strippingTrailingSlash
        guard let url = URL(string: "\(base)/bug-reports") else {
            throw BugReportError.invalidURL
        }

        // Build multipart body manually — Foundation has no built-in helper.
        let boundary = "----DraftRightBugReport-\(UUID().uuidString)"
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        request.setValue("multipart/form-data; boundary=\(boundary)",
                         forHTTPHeaderField: "Content-Type")
        if let token = authToken, !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let appVersion: String = {
            let info = Bundle.main.infoDictionary
            let short = info?["CFBundleShortVersionString"] as? String ?? "?"
            let build = info?["CFBundleVersion"] as? String ?? "?"
            return "\(short)+\(build)"
        }()
        let osInfo = ProcessInfo.processInfo.operatingSystemVersionString

        let context: [String: Any] = [
            "platform": "macos",
            "locale": Locale.current.identifier,
            "ts": ISO8601DateFormatter().string(from: Date())
        ]
        let contextJSON = (try? JSONSerialization.data(withJSONObject: context))
            .flatMap { String(data: $0, encoding: .utf8) } ?? "{}"

        var body = Data()
        body.appendFormField(name: "description", value: trimmed, boundary: boundary)
        body.appendFormField(name: "source", value: source, boundary: boundary)
        body.appendFormField(name: "app_version", value: appVersion, boundary: boundary)
        body.appendFormField(name: "os_info", value: osInfo, boundary: boundary)
        body.appendFormField(name: "context", value: contextJSON, boundary: boundary)
        if let email = userEmail?.trimmingCharacters(in: .whitespacesAndNewlines),
           !email.isEmpty {
            body.appendFormField(name: "user_email", value: email, boundary: boundary)
        }
        if let shot = screenshot {
            let mime = mimeType(forFilename: shot.filename)
            body.appendFileField(
                name: "screenshot",
                filename: shot.filename,
                mimeType: mime,
                fileData: shot.data,
                boundary: boundary
            )
        }
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)
        request.httpBody = body

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            throw BugReportError.network(error)
        }

        guard let http = response as? HTTPURLResponse else {
            throw BugReportError.emptyResponse
        }
        if http.statusCode >= 400 {
            let bodyText = String(data: data, encoding: .utf8) ?? ""
            throw BugReportError.httpError(http.statusCode, bodyText)
        }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let idString = json["id"] as? String,
              let uuid = UUID(uuidString: idString) else {
            throw BugReportError.emptyResponse
        }
        return uuid
    }

    private static func mimeType(forFilename filename: String) -> String {
        let lower = filename.lowercased()
        if lower.hasSuffix(".png") { return "image/png" }
        if lower.hasSuffix(".jpg") || lower.hasSuffix(".jpeg") { return "image/jpeg" }
        return "application/octet-stream"
    }
}

private extension Data {
    mutating func appendFormField(name: String, value: String, boundary: String) {
        append("--\(boundary)\r\n".data(using: .utf8)!)
        append("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n"
            .data(using: .utf8)!)
        append("\(value)\r\n".data(using: .utf8)!)
    }

    mutating func appendFileField(
        name: String,
        filename: String,
        mimeType: String,
        fileData: Data,
        boundary: String
    ) {
        append("--\(boundary)\r\n".data(using: .utf8)!)
        append("Content-Disposition: form-data; name=\"\(name)\"; filename=\"\(filename)\"\r\n"
            .data(using: .utf8)!)
        append("Content-Type: \(mimeType)\r\n\r\n".data(using: .utf8)!)
        append(fileData)
        append("\r\n".data(using: .utf8)!)
    }
}
